wgpu-core/src/pipeline.rs

```diff
@@ -218,6 +229,32 @@ impl CreateShaderModuleError {
             _ => None,
         }
     }
+
+    pub fn source_range(&self) -> Option<std::ops::Range<usize>> {
+        match self {
+            #[cfg(feature = "wgsl")]
+            CreateShaderModuleError::Parsing(err) => err
+                .inner
+                .labels()
+                .next()
+                .and_then(|(span, _)| span.to_range()),
+            #[cfg(feature = "glsl")]
+            CreateShaderModuleError::ParsingGlsl(err) => {
+                err.inner.errors.first().and_then(|e| e.meta.to_range())
```

Comment:
This one feels odd, most of the error types report one error. Meanwhile ParsingGlsl seems to be able to return multiple errors?

Reply:
well the others are also returning iterators and you're only calling `next()` on them, thus only looking at the first one. All of them have several error spans.

cc: @teoxoy generally we should expose these errors one by one to webgpu right?

Reply:
Right now only the GLSL frontend reports multiple errors but the WGSL frontend will as well in the future.

WGSL errors are richer and contain multiple spans so that we can point to different parts of the shader, see for example the redefinition error:

https://github.com/gfx-rs/wgpu/blob/82fa5801525e7a32118ba3e02b9ce4adfb57112a/naga/src/front/wgsl/error.rs#L610

Reply:
Is this new method necessary? It seems to be doing the same thing as the one above.

Reply:
> WGSL errors are richer and contain multiple spans so that we can point to different parts of the shader, see for example the redefinition error:

Thank you for the example! How do I best report that as a WebGPU error?
I'm basically struggling with the fact that the official WebGPU `GPUCompilationMessage` only lets one include a single location
https://developer.mozilla.org/en-US/docs/Web/API/GPUCompilationMessage

One could
- add a `Vec<>` for the compilation spans to the wgpu `CompilationMessage` type, deviating from the standard
- open an issue in the repository where the work on the WebGPU specification happens, and making the case for "multiple locations make sense"
- report multiple spans as multiple errors (e.g. Make the first one an error, and the other ones warnings that pretty much say "error also happened here")



> Is this new method necessary? It seems to be doing the same thing as the one above.

I added a separate method, since the existing location method requires a source string, and it computes a few pieces of information that I just end up throwing away. (line number and offset in the line)

I'd be happy about better suggestions though. Maybe I could tweak the existing `location` methods to return a `Range` and defer resolving the `naga::SourceLocation` until a bit later?

Reply:
> How do I best report that as a WebGPU error?
I'm basically struggling with the fact that the official WebGPU `GPUCompilationMessage` only lets one include a single location

I think a single location is fine, as for the error message we can use the `ParseError.emit_to_string()` method (it will return a nicely formatted error message that takes into account all the spans).

The implementation of `.location()` seems to use the first span which I think is correct? for now? I guess it makes sense that the first span would point to the issue at hand and the others are complementary but it doesn't seem to be documented.

> I added a separate method, since the existing location method requires a source string, and it computes a few pieces of information that I just end up throwing away. (line number and offset in the line)

Why do we need to throw some of the fields away? All `naga::SourceLocation`'s fields are part of `GPUCompilationMessage`.

Reply:
Sounds good.

Designing a good `CompilationMessage` for wgpu is a bit annoying, because we pretty much have to deviate from [the standard](https://gpuweb.github.io/gpuweb/#gpucompilationmessage) for it to be idiomatic Rust. Here's what I'm leaning towards at the moment, does this make sense?
```

pub struct CompilationMessage {
    pub message: String,
    // `message_type` is a name that doesn't clash with a reserved keyword, unlike `type`
    pub message_type: CompilationMessageType,
    /// An option, instead of the "Must be 0 if the message does not correspond to any specific point in the shader code." rule. That rule is a tad wild.
    pub location: Option<SourceLocation>,
}

// Exists so that we can put the location info in an option
pub struct SourceLocation {
    /// Assumes a UTF-8 encoded source instead of assuming UTF-16
    pub offset: u32,
    /// Assumes a UTF-8 encoded source instead of assuming UTF-16
    pub length: u32,
    /// This one actually matches the standard. Well, except that the name is `line_number` instead of `line_num`.
    pub line_number: u32,
    /// Uses Unicode code units, which are more comprehensible for humans instead of UTF-16 offsets.
    /// Naga does the same thing, hence this decision. I'm not sure if it's ideal though. 
    /// e.g. VSCode's column indicator actually uses UTF-16 to count, as far as I can tell. And Unicode code units also aren't perfect for humans, since they don't match up with visual characters (graphemes) 
    pub line_position: u32,
}
```

Would love some feedback.